File: jarvis/server/src/tools/sandbox.ts
/**
 * Sandboxed code execution: exec_js and exec_shell.
 *
 * `exec_js` runs arbitrary JavaScript in a throwaway `fork`ed Node
 * subprocess with a 5s timeout and a temp cwd, after a source-level
 * denylist rejects known-dangerous constructs. This is defense-in-depth,
 * not a real sandbox (no seccomp/chroot/network namespace) — the denylist
 * plus a short timeout and a scratch cwd are the guarantees actually
 * provided, mirroring the Python edition's AST-check approach.
 *
 * `exec_shell` runs a single allowlisted command with `execFile` (no shell
 * interpretation at all), so shell metacharacters are inert.
 */

import { fork, execFile } from 'node:child_process';
import { promises as fs } from 'node:fs';
import os from 'node:os';
import path from 'node:path';

import { registry } from './base.js';
import type { Tool, ToolResult } from '../types.js';

const JS_TIMEOUT_MS = 5_000;
const SHELL_TIMEOUT_MS = 5_000;

// Modules that are a direct route to the network, process control, or
// dynamic code loading — never allowed, regardless of import style.
const BANNED_MODULES = [
  'child_process', 'node:child_process',
  'net', 'node:net',
  'http', 'node:http',
  'https', 'node:https',
  'http2', 'node:http2',
  'dgram', 'node:dgram',
  'tls', 'node:tls',
  'cluster', 'node:cluster',
  'worker_threads', 'node:worker_threads',
  'vm', 'node:vm',
  'inspector', 'node:inspector',
  'repl', 'node:repl',
];

/** Matches `require('x')`, `require("x")`, `import ... from 'x'`, `import('x')`. */
function findModuleReferences(source: string): string[] {
  const found: string[] = [];
  const patterns = [
    /require\(\s*(['"])(.*?)\1\s*\)/g,
    /import\s+(?:[\s\S]*?\s+from\s+)?(['"])(.*?)\1/g,
    /import\(\s*(['"])(.*?)\1\s*\)/g,
  ];
  for (const re of patterns) {
    for (const match of source.matchAll(re)) {
      found.push(match[2]);
    }
  }
  return found;
}

/** `fs.writeFile*`/`appendFile*`/`unlink*`/`rm*` calls whose literal path arg looks like it escapes the sandbox cwd. */
function findUnsafeFsWrites(source: string): string[] {
  const violations: string[] = [];
  const writeCallRe = /\b(?:fs\.)?(writeFileSync|writeFile|appendFileSync|appendFile|unlinkSync|unlink|rmSync|rmdirSync|rmdir|rm)\s*\(\s*(['"])(.*?)\2/g;
  for (const match of source.matchAll(writeCallRe)) {
    const [, fn, , arg] = match;
    if (arg.startsWith('/') || arg.startsWith('~') || arg.includes('..')) {
      violations.push(`write-mode ${fn}() outside sandbox cwd: ${JSON.stringify(arg)}`);
    }
  }
  return violations;
}

/** Parse `source` and return a list of denylist violation messages. */
export function checkJsSource(source: string): string[] {
  const violations: string[] = [];

  for (const mod of findModuleReferences(source)) {
    const root = mod.split('/')[0];
    if (BANNED_MODULES.includes(mod) || BANNED_MODULES.includes(root)) {
      violations.push(`import of banned module: ${mod}`);
    }
  }

  // Escapes to dynamic module loading / eval-style code execution.
  if (/\brequire\s*\(\s*[^'"]/.test(source)) {
    violations.push('call to require() with a non-literal argument (dynamic require escape)');
  }
  if (/\bimport\s*\(\s*[^'")]/.test(source)) {
    violations.push('call to dynamic import() with a non-literal argument');
  }
  if (/\beval\s*\(/.test(source)) {
    violations.push('call to banned builtin: eval');
  }
  if (/\bnew\s+Function\s*\(/.test(source)) {
    violations.push('call to banned builtin: Function constructor');
  }
  if (/\bprocess\s*\.\s*exit\s*\(/.test(source)) {
    violations.push('call to banned attribute: process.exit');
  }
  if (/\bprocess\s*\.\s*kill\s*\(/.test(source)) {
    violations.push('call to banned attribute: process.kill');
  }
  if (/\bprocess\s*\.\s*binding\s*\(/.test(source)) {
    violations.push('call to banned attribute: process.binding');
  }
  if (/\bglobalThis\s*\[\s*['"]process['"]\s*\]/.test(source)) {
    violations.push('indirect access to process via globalThis indexing');
  }

  violations.push(...findUnsafeFsWrites(source));

  return violations;
}

function fail(summary: string): ToolResult {
  return { ok: false, output: '', summary };
}

const execJsTool: Tool = {
  name: 'exec_js',
  description:
    'Execute a short JavaScript snippet in an isolated Node subprocess with a 5s timeout and a temp ' +
    'working directory. Network/process/require-escape calls are statically rejected before execution.',
  parameters: {
    type: 'object',
    properties: {
      code: { type: 'string', minLength: 1, description: 'JavaScript source to execute.' },
    },
    required: ['code'],
    additionalProperties: false,
  },
  async run(args): Promise<ToolResult> {
    const code = args.code as string;

    const violations = checkJsSource(code);
    if (violations.length > 0) {
      return fail(`Rejected by sandbox denylist: ${violations.join('; ')}`);
    }

    const tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'jarvis-sandbox-'));
    const scriptPath = path.join(tmpDir, 'snippet.js');
    try {
      await fs.writeFile(scriptPath, code, 'utf-8');

      const result = await new Promise<ToolResult>((resolve) => {
        const stdoutChunks: Buffer[] = [];
        const stderrChunks: Buffer[] = [];

        const child = fork(scriptPath, [], {
          cwd: tmpDir,
          stdio: ['ignore', 'pipe', 'pipe', 'ipc'],
          // No inherited env leakage beyond what the child needs to run node.
          env: { PATH: process.env.PATH ?? '' },
          execArgv: [], // no --inspect / loader escapes
        });

        let settled = false;
        const timer = setTimeout(() => {
          if (settled) return;
          settled = true;
          child.kill('SIGKILL');
          resolve(fail(`exec_js timed out after ${JS_TIMEOUT_MS / 1000}s`));
        }, JS_TIMEOUT_MS);

        child.stdout?.on('data', (chunk: Buffer) => stdoutChunks.push(chunk));
        child.stderr?.on('data', (chunk: Buffer) => stderrChunks.push(chunk));

        child.on('error', (err) => {
          if (settled) return;
          settled = true;
          clearTimeout(timer);
          resolve(fail(`exec_js failed to start: ${err.message}`));
        });

        child.on('exit', (code2) => {
          if (settled) return;
          settled = true;
          clearTimeout(timer);
          const outText = Buffer.concat(stdoutChunks).toString('utf-8');
          const errText = Buffer.concat(stderrChunks).toString('utf-8');
          const ok = code2 === 0;
          const combined = ok ? outText : `${outText}\n${errText}`.trim();
          resolve({
            ok,
            output: combined,
            summary: ok ? 'exec_js succeeded' : `exec_js exited ${code2}`,
            meta: { returncode: code2 },
          });
        });
      });

      return result;
    } finally {
      await fs.rm(tmpDir, { recursive: true, force: true }).catch(() => undefined);
    }
  },
};

// Commands considered harmless enough to run with zero side effects beyond
// reading already-public system state.
const SHELL_ALLOWLIST = new Set(['ls', 'cat', 'date', 'uname', 'df', 'ps', 'echo', 'pwd', 'which']);

/** Minimal shell-style tokenizer supporting quoted strings — no metacharacter interpretation follows. */
function tokenize(command: string): string[] | null {
  const tokens: string[] = [];
  let i = 0;
  const n = command.length;
  while (i < n) {
    while (i < n && /\s/.test(command[i])) i++;
    if (i >= n) break;
    let token = '';
    let quote: '"' | "'" | null = null;
    while (i < n) {
      const ch = command[i];
      if (quote) {
        if (ch === quote) {
          quote = null;
          i++;
        } else {
          token += ch;
          i++;
        }
      } else if (ch === '"' || ch === "'") {
        quote = ch;
        i++;
      } else if (/\s/.test(ch)) {
        break;
      } else {
        token += ch;
        i++;
      }
    }
    if (quote) return null; // unterminated quote
    tokens.push(token);
  }
  return tokens;
}

const execShellTool: Tool = {
  name: 'exec_shell',
  description: 'Run a single allowlisted read-only shell command (ls, cat, date, uname, df, ps, echo, pwd, which).',
  parameters: {
    type: 'object',
    properties: {
      command: { type: 'string', minLength: 1, description: "Full command line, e.g. 'ls -la'." },
    },
    required: ['command'],
    additionalProperties: false,
  },
  async run(args): Promise<ToolResult> {
    const command = args.command as string;
    const tokens = tokenize(command);
    if (tokens === null) {
      return fail('Could not parse command: unterminated quote');
    }
    if (tokens.length === 0) {
      return fail('Empty command');
    }

    const [program, ...rest] = tokens;
    if (!SHELL_ALLOWLIST.has(program)) {
      return fail(`Command not allowlisted: ${JSON.stringify(program)}`);
    }

    return new Promise<ToolResult>((resolve) => {
      // execFile with an argument array — shell=false by construction, so
      // shell metacharacters in `rest` are inert, never interpreted.
      const child = execFile(
        program,
        rest,
        { timeout: SHELL_TIMEOUT_MS, windowsHide: true },
        (error, stdout, stderr) => {
          if (error && (error as NodeJS.ErrnoException).code === 'ENOENT') {
            resolve(fail(`Command not found: ${JSON.stringify(program)}`));
            return;
          }
          if (error && error.killed) {
            resolve(fail(`exec_shell timed out after ${SHELL_TIMEOUT_MS / 1000}s`));
            return;
          }
          const ok = !error;
          const returncode = ok ? 0 : typeof error?.code === 'number' ? error.code : 1;
          const combined = ok ? stdout : `${stdout}\n${stderr}`.trim();
          resolve({
            ok,
            output: combined,
            summary: `${program} exited ${returncode}`,
            meta: { returncode },
          });
        },
      );
      void child;
    });
  },
};

registry.register(execJsTool);
registry.register(execShellTool);

File: jarvis/backend/app/core/tools/sandbox.py
"""Sandboxed code execution: exec_python and exec_shell.

`exec_python` runs arbitrary Python in a throwaway subprocess with a 5s
timeout and a temp cwd, after an AST-level static check rejects known
dangerous constructs. This is defense-in-depth, not a real sandbox (no
seccomp/chroot/network namespace) — the AST denylist plus a short timeout
and a scratch cwd are the guarantees actually provided.

`exec_shell` runs a single allowlisted command with no shell interpretation
at all (`shell=False`), so shell metacharacters are inert.
"""

from __future__ import annotations

import ast
import asyncio
import shlex
import tempfile
from pathlib import Path
from typing import Any

from app.core.tools.base import register
from app.models.schemas import ToolResult

PYTHON_TIMEOUT_S = 5.0
SHELL_TIMEOUT_S = 5.0

# Whole modules that are never allowed to be imported by sandboxed code —
# each one is a direct route to the network, process control, or bulk
# filesystem destruction.
_BANNED_MODULES = {"subprocess", "socket", "shutil", "ctypes", "importlib", "multiprocessing", "os.path"}

# Dotted attribute paths banned even if the owning module is otherwise fine
# (e.g. `os` itself is allowed for things like `os.getcwd()`).
_BANNED_ATTR_CHAINS = {
    ("os", "system"), ("os", "popen"), ("os", "exec"), ("os", "execv"), ("os", "execve"),
    ("os", "execvp"), ("os", "spawn"), ("os", "spawnl"), ("os", "spawnv"), ("os", "fork"),
    ("os", "kill"), ("os", "remove"), ("os", "unlink"), ("os", "rmdir"),
}

# Bare names that are dangerous regardless of attribute access — these cover
# the classic __import__-tricks path to reaching banned modules dynamically.
_BANNED_CALL_NAMES = {"__import__", "eval", "exec", "compile", "globals", "vars"}


class SandboxViolation(Exception):
    """Raised by the AST checker when code contains a denylisted construct."""


class _SafetyVisitor(ast.NodeVisitor):
    """Walks the AST once, collecting every denylist violation it finds."""

    def __init__(self) -> None:
        self.violations: list[str] = []

    def visit_Import(self, node: ast.Import) -> None:
        for alias in node.names:
            root = alias.name.split(".")[0]
            if alias.name in _BANNED_MODULES or root in _BANNED_MODULES:
                self.violations.append(f"import of banned module: {alias.name}")
        self.generic_visit(node)

    def visit_ImportFrom(self, node: ast.ImportFrom) -> None:
        module = node.module or ""
        root = module.split(".")[0]
        if module in _BANNED_MODULES or root in _BANNED_MODULES:
            self.violations.append(f"import from banned module: {module}")
        self.generic_visit(node)

    def visit_Call(self, node: ast.Call) -> None:
        # Bare dangerous builtins: __import__("os").system(...) etc.
        if isinstance(node.func, ast.Name) and node.func.id in _BANNED_CALL_NAMES:
            self.violations.append(f"call to banned builtin: {node.func.id}")

        # Dotted attribute calls like os.system(...) or shutil.rmtree(...).
        if isinstance(node.func, ast.Attribute) and isinstance(node.func.value, ast.Name):
            chain = (node.func.value.id, node.func.attr)
            if chain in _BANNED_ATTR_CHAINS:
                self.violations.append(f"call to banned attribute: {chain[0]}.{chain[1]}")
            if chain == ("shutil", "rmtree"):
                self.violations.append("call to banned attribute: shutil.rmtree")

        # open(path, mode) in write/append/exclusive/plus mode with a path
        # literal that looks like it targets outside the scratch cwd.
        if isinstance(node.func, ast.Name) and node.func.id == "open":
            self._check_open_call(node)

        self.generic_visit(node)

    def _check_open_call(self, node: ast.Call) -> None:
        mode = ""
        if len(node.args) >= 2 and isinstance(node.args[1], ast.Constant):
            mode = str(node.args[1].value)
        for kw in node.keywords:
            if kw.arg == "mode" and isinstance(kw.value, ast.Constant):
                mode = str(kw.value.value)
        is_write = any(c in mode for c in ("w", "a", "x", "+"))
        if not is_write:
            return
        if node.args and isinstance(node.args[0], ast.Constant) and isinstance(node.args[0].value, str):
            path = node.args[0].value
            if path.startswith("/") or path.startswith("~") or ".." in path:
                self.violations.append(f"write-mode open() outside sandbox cwd: {path!r}")


def check_python_source(source: str) -> list[str]:
    """Parse `source` and return a list of denylist violation messages."""
    try:
        tree = ast.parse(source)
    except SyntaxError as exc:
        return [f"syntax error: {exc}"]
    visitor = _SafetyVisitor()
    visitor.visit(tree)
    return visitor.violations


@register
class ExecPythonTool:
    name = "exec_python"
    description = (
        "Execute a short Python snippet in an isolated subprocess with a 5s "
        "timeout and a temp working directory. Network/process/bulk-delete "
        "calls are statically rejected before execution."
    )
    schema: dict[str, Any] = {
        "type": "object",
        "properties": {
            "code": {"type": "string", "minLength": 1, "description": "Python source to execute."},
        },
        "required": ["code"],
        "additionalProperties": False,
    }

    async def run(self, code: str) -> ToolResult:
        violations = check_python_source(code)
        if violations:
            return ToolResult(
                ok=False,
                output="",
                summary=f"Rejected by sandbox denylist: {'; '.join(violations)}",
            )

        with tempfile.TemporaryDirectory(prefix="jarvis-sandbox-") as tmp_dir:
            script_path = Path(tmp_dir) / "snippet.py"
            script_path.write_text(code, encoding="utf-8")

            try:
                proc = await asyncio.create_subprocess_exec(
                    "python3",
                    str(script_path),
                    cwd=tmp_dir,
                    stdout=asyncio.subprocess.PIPE,
                    stderr=asyncio.subprocess.PIPE,
                )
                stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=PYTHON_TIMEOUT_S)
            except asyncio.TimeoutError:
                proc.kill()
                await proc.wait()
                return ToolResult(ok=False, output="", summary=f"exec_python timed out after {PYTHON_TIMEOUT_S}s")

            out_text = stdout.decode("utf-8", errors="replace")
            err_text = stderr.decode("utf-8", errors="replace")
            ok = proc.returncode == 0
            combined = out_text if ok else f"{out_text}\n{err_text}".strip()
            summary = "exec_python succeeded" if ok else f"exec_python exited {proc.returncode}"
            return ToolResult(ok=ok, output=combined, summary=summary, meta={"returncode": proc.returncode})


# Commands considered harmless enough to run with zero arguments-side-effects
# beyond reading already-public system state.
_SHELL_ALLOWLIST = {"ls", "cat", "date", "uname", "df", "ps", "echo", "pwd", "which"}


@register
class ExecShellTool:
    name = "exec_shell"
    description = "Run a single allowlisted read-only shell command (ls, cat, date, uname, df, ps, echo, pwd, which)."
    schema: dict[str, Any] = {
        "type": "object",
        "properties": {
            "command": {"type": "string", "minLength": 1, "description": "Full command line, e.g. 'ls -la'."},
        },
        "required": ["command"],
        "additionalProperties": False,
    }

    async def run(self, command: str) -> ToolResult:
        try:
            tokens = shlex.split(command)
        except ValueError as exc:
            return ToolResult(ok=False, output="", summary=f"Could not parse command: {exc}")

        if not tokens:
            return ToolResult(ok=False, output="", summary="Empty command")

        program = tokens[0]
        if program not in _SHELL_ALLOWLIST:
            return ToolResult(ok=False, output="", summary=f"Command not allowlisted: {program!r}")

        try:
            proc = await asyncio.create_subprocess_exec(
                *tokens,  # shell=False by construction — no metachar interpretation
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=SHELL_TIMEOUT_S)
        except asyncio.TimeoutError:
            proc.kill()
            await proc.wait()
            return ToolResult(ok=False, output="", summary=f"exec_shell timed out after {SHELL_TIMEOUT_S}s")
        except FileNotFoundError:
            return ToolResult(ok=False, output="", summary=f"Command not found: {program!r}")

        out_text = stdout.decode("utf-8", errors="replace")
        err_text = stderr.decode("utf-8", errors="replace")
        ok = proc.returncode == 0
        combined = out_text if ok else f"{out_text}\n{err_text}".strip()
        return ToolResult(
            ok=ok,
            output=combined,
            summary=f"{program} exited {proc.returncode}",
            meta={"returncode": proc.returncode},
        )
